action.go

```diff
@@ -155,48 +45,55 @@ func (c *client) do(table, row []byte, action action, useCache bool, retries int
 			Region: regionSpecifier,
 			Call:   a.ToProto().(*proto.CoprocessorServiceCall),
 		})
+	default:
+		return nil, errors.Errorf("Unknown action - %T - %v", action, action)
 	}
 
-	result := make(chan pb.Message)
-
-	go func() {
-		r := <-cl.responseCh
+	err = conn.call(cl)
+	if err != nil {
+		delete(c.cachedConns, region.Server)
+		return nil, errors.Trace(err)
+	}
 
-		switch r.(type) {
-		case *exception:
-			if retries <= c.maxRetries {
-				// retry action, and refresh region info
-				log.Warnf("Retrying action for the %d time(s)", retries+1)
-				// clean old region info
-				c.CleanRegionCache(table)
-				RetrySleep(retries + 1)
-				newr := c.do(table, row, action, false, retries+1)
-				result <- <-newr
-			} else {
-				result <- r
-			}
-			return
-		default:
-			result <- r
-		}
-	}()
+	return cl, nil
+}
 
-	if cl != nil {
-		err := conn.call(cl)
+func (c *client) innerDo(table, row []byte, action action, useCache bool) (pb.Message, error) {
+	// Try to create and send a new resuqest call.
+	cl, err := c.innerCall(table, row, action, useCache)
+	if err != nil {
+		log.Warnf("inner call failed - %v", errors.ErrorStack(err))
+		// remove bad server cache and try again
```

Comment:
Line 87 in do function,  in old code i see clean cache only when get a exception response from server.